src/cpuid/src/lib.rs

```diff
@@ -5,68 +5,693 @@
+                brand_string[41],
+                brand_string[42],
+                brand_string[43],
+            ]);
+            leaf.result.edx = u32::from_ne_bytes([
+                brand_string[44],
+                brand_string[45],
+                brand_string[46],
+                brand_string[47],
+            ]);
+        }
+
+        Ok(())
+    }
+}
+
+impl CpuidTrait for Cpuid {
+    /// Gets a given sub-leaf.
+    #[inline]
+    fn get(&self, key: &CpuidKey) -> Option<&CpuidEntry> {
+        match self {
+            Self::Intel(intel_cpuid) => intel_cpuid.get(key),
+            Self::Amd(amd_cpuid) => amd_cpuid.get(key),
+        }
+    }
+
+    /// Gets a given sub-leaf.
+    #[inline]
+    fn get_mut(&mut self, key: &CpuidKey) -> Option<&mut CpuidEntry> {
+        match self {
+            Self::Intel(intel_cpuid) => intel_cpuid.get_mut(key),
+            Self::Amd(amd_cpuid) => amd_cpuid.get_mut(key),
+        }
+    }
+}
+
+impl Cpuid {
+    /// Gets supported CPUID by KVM.
+    ///
+    /// # Errors
+    ///
+    /// When failed to access KVM.
+    #[cfg(cpuid)]
+    #[inline]
+    pub fn kvm_get_supported_cpuid() -> std::result::Result<Self, KvmGetSupportedCpuidError> {
+        let supported_kvm_cpuid =
+            kvm_ioctls::Kvm::new()?.get_supported_cpuid(kvm_bindings::KVM_MAX_CPUID_ENTRIES)?;
+        let supported_raw_cpuid = RawCpuid::from(supported_kvm_cpuid);
+        Cpuid::try_from(supported_raw_cpuid).map_err(KvmGetSupportedCpuidError::CpuidFromRaw)
+    }
+    /// Returns `Some(&IntelCpuid)` if `Self == Self::Intel(_)` else returns `None`.
+    #[inline]
+    #[must_use]
+    pub fn intel_mut(&mut self) -> Option<&mut IntelCpuid> {
+        match self {
+            Self::Intel(intel) => Some(intel),
+            Self::Amd(_) => None,
+        }
+    }
+    /// Returns `Some(&mut IntelCpuid)` if `Self == Self::Intel(_)` else returns `None`.
+    #[inline]
+    #[must_use]
+    pub fn intel(&self) -> Option<&IntelCpuid> {
+        match self {
+            Self::Intel(intel) => Some(intel),
+            Self::Amd(_) => None,
+        }
+    }
+    /// Returns `Some(&AmdCpuid)` if `Self == Self::Amd(_)` else returns `None`.
+    #[inline]
+    #[must_use]
+    pub fn amd(&self) -> Option<&AmdCpuid> {
+        match self {
+            Self::Intel(_) => None,
+            Self::Amd(amd) => Some(amd),
+        }
+    }
+    /// Returns `Some(&mut AmdCpuid)` if `Self == Self::Amd(_)` else returns `None`.
+    #[inline]
+    #[must_use]
+    pub fn amd_mut(&mut self) -> Option<&mut AmdCpuid> {
+        match self {
+            Self::Intel(_) => None,
+            Self::Amd(amd) => Some(amd),
+        }
+    }
+    /// Applies required modifications to CPUID respective of a vCPU.
+    ///
+    /// # Errors
+    ///
+    /// When:
+    /// - [`IntelCpuid::normalize`] errors.
+    /// - [`AmdCpuid::normalize`] errors.
+    // As we pass through host frequency, we require CPUID and thus `cfg(cpuid)`.
```

Comment:
```
// As we pass through host frequency, we require CPUID and thus `cfg(cpuid)`.
```
Is the justification for having `cfg(cpuid)` on the respective item. It is required since we need to make a call to `cpuid` via [`core::arch::x86_64::__cpuid_count`](https://doc.rust-lang.org/core/arch/x86_64/fn.__cpuid_count.html) and we can only do this safely when `cpuid` is supported and we consider `cpuid` supported depending on `cfg(cpuid)`.